cmake/git-versioning.cmake

```diff
@@ -0,0 +1,124 @@
+
+        if (${PROJECT_NAME}_git_status MATCHES "changed")
+            set(${PROJECT_NAME}_git_status "dirty")
+        else ()
+            unset(${PROJECT_NAME}_git_status)
+        endif ()
+
+        # Get version components
+        set(semver_normal_regex "v(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)")
+        set(semver_prerelease_regex "(-((0|[1-9][0-9]*|[0-9]*[a-zA-Z-][0-9a-zA-Z-]*)(\\.(0|[1-9][0-9]*|[0-9]*[a-zA-Z-][0-9a-zA-Z-]*))*))")
+        set(semver_regex "^${semver_normal_regex}${semver_prerelease_regex}?$")
+        if (NOT ${PROJECT_NAME}_git_tag MATCHES ${semver_regex})
+            message(FATAL_ERROR "Tag, \"${${PROJECT_NAME}_git_tag}\", does not conform to SemVer \"${${PROJECT_NAME}_semver_regex}\".")
+        endif ()
+
+        string(REGEX REPLACE ${semver_regex} "\\1" ${PROJECT_NAME}_version_major "${${PROJECT_NAME}_git_tag}")
+        string(REGEX REPLACE ${semver_regex} "\\2" ${PROJECT_NAME}_version_minor "${${PROJECT_NAME}_git_tag}")
+        string(REGEX REPLACE ${semver_regex} "\\3" ${PROJECT_NAME}_version_patch "${${PROJECT_NAME}_git_tag}")
+        if (${PROJECT_NAME}_git_tag MATCHES "^${semver_normal_regex}${semver_prerelease_regex}$")
+            string(REGEX REPLACE ${semver_regex} "\\4" ${PROJECT_NAME}_version_prerelease "${${PROJECT_NAME}_git_tag}")
+        else ()
+            set(${PROJECT_NAME}_version_prerelease "")
+        endif ()
+
+        # Build meta info
+        if (NOT ${PROJECT_NAME}_git_build_number MATCHES "^0$")
+            set(${PROJECT_NAME}_version_meta "+${${PROJECT_NAME}_git_branch}.${${PROJECT_NAME}_git_sha}.${${PROJECT_NAME}_git_build_number}")
+            if (DEFINED ${PROJECT_NAME}_git_status)
+                set(${PROJECT_NAME}_version_meta "${${PROJECT_NAME}_version_meta}.${${PROJECT_NAME}_git_status}")
+            endif ()
+        else ()
+            if (DEFINED ${PROJECT_NAME}_git_status)
+                set(${PROJECT_NAME}_version_meta "+${${PROJECT_NAME}_git_status}")
+            else ()
+                set(${PROJECT_NAME}_version_meta "")
+            endif ()
+        endif ()
+
+
+    else ()
+
+        message(WARNING "Failed to find git executable. Unable to establish version information for ${PROJECT_NAME}.")
+        set(${PROJECT_NAME}_version_major "0")
+        set(${PROJECT_NAME}_version_minor "0")
+        set(${PROJECT_NAME}_version_patch "0")
+        set(${PROJECT_NAME}_version_meta "+git.not.found")
+
+    endif ()
+
+    set(${PROJECT_NAME}_version "v${${PROJECT_NAME}_version_major}.${${PROJECT_NAME}_version_minor}.${${PROJECT_NAME}_version_patch}${${PROJECT_NAME}_version_prerelease}${${PROJECT_NAME}_version_meta}")
+
+endmacro()
+
+macro(add_version_variable version_variable)
+    string(TOUPPER ${version_variable} temp_upper)
+    string(APPEND content "#define ${temp_upper} ${${version_variable}}\n")
+endmacro()
+
+macro(make_version_header)
+
+    set(content "#pragma once\n\n")
```

Comment:
It sounds like this is a bigger discussion and, practically speaking, seems like this just works but I wonder if we should be using `#pragma once` vs include guards?

Reply:
Yes. We should decide on this relatively soon. My big gripe about include guards is the repetitive nature of them. If someone were to create a tool that automatically applied them, I wouldn't worry as much about it. 😉

Reply:
I'm not sure what the current status of #pragma once is, and I know it's supported by multiple preprocessors now, but my experience is that `#pragma once` is not path-agnostic - so if you have a "common.h" in your project folder and also a "common.h" in some submodule or sister folder, and they both use `#pragma once `instead of include guards, they will both be included and potentially clash. Include guards only allow the symbol to be defined once, and the preprocessor should complain if you have two of the same file name.
I'd have to start a test project to see if this is still true.